File: 0360 322 Object-Oriented Software Analysis and Design/Assignments/A3/src/CollidesWithVisitor.java
public class CollidesWithVisitor extends EntityVisitor {

    // second argument
    private Entity secondEntity;

    CollidesWithVisitor(Entity secondEntity) {          // construct, remember second argument
        this.secondEntity = secondEntity;
    }

    public void visit(Asteroid asteroid) {              // asteroid colliding with a ...??
        secondEntity.doCollision(asteroid);             // determine what secondEntity is
    }
    public void visit(Spaceship spaceship) {            // spaceship colliding with a ...??
        secondEntity.doCollision(spaceship);            // determine what secondEntity is
    }
    public void visit(Moon moon) {                      // moon colliding with a ...??
        secondEntity.doCollision(moon);                 // determine what secondEntity is
    }
}

File: 0360 322 Object-Oriented Software Analysis and Design/Assignments/DynamicDispatch/DynamicDispatch/CollidesWithVisitor.cs


namespace A3Test
{
    public class CollidesWithVisitor : IVisitor
    {

        private readonly SpaceEntity _secondEntity;

        public CollidesWithVisitor(SpaceEntity secondEntity) => _secondEntity = secondEntity;

        public void Visit(Asteroid astereoid) => Interaction.Collide(_secondEntity, astereoid);

        public void Visit(Spaceship spaceship) => Interaction.Collide(_secondEntity, spaceship);

        public void Visit(Moon moon) => Interaction.Collide(_secondEntity, moon);
    }
}
